server/src/main/java/io/crate/execution/engine/sort/Sort.java

```diff
@@ -22,48 +22,78 @@
+import static org.apache.lucene.util.RamUsageEstimator.NUM_BYTES_OBJECT_REF;
 
 import java.util.ArrayList;
 import java.util.Comparator;
 import java.util.List;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.Executor;
+import java.util.function.LongConsumer;
+
+
+import org.apache.lucene.util.RamUsageEstimator;
 
 import io.crate.common.collections.Iterables;
 import io.crate.common.collections.Lists;
 
 public final class Sort {
 
+    @SuppressWarnings("unchecked")
     public static <T> CompletableFuture<List<T>> parallelSort(List<T> list,
+                                                              LongConsumer allocateBytes,
                                                               Comparator<? super T> comparator,
                                                               int minItemsPerThread,
                                                               int numAvailableThreads,
                                                               Executor executor) {
 
         int itemsPerThread = list.size() / Math.max(1, numAvailableThreads);
-        if (numAvailableThreads <= 1 || itemsPerThread < minItemsPerThread) {
-            list.sort(comparator);
-            return completedFuture(list);
-        }
+//        if (numAvailableThreads <= 1 || itemsPerThread < minItemsPerThread) {
+//            list.sort(comparator);
+//            allocateBytes.accept(sortOverheadBytes(list.size()));
+//            return completedFuture(list);
+//        }
+        //89 -> sublist 96-97 -> 99 00, 01 02
         List<List<T>> partitions = Lists.partition(list, itemsPerThread);
-        ArrayList<CompletableFuture<List<T>>> futures = new ArrayList<>(partitions.size());
+        CompletableFuture<List<T>>[] futures = new CompletableFuture[partitions.size()];
+        allocateBytes.accept(RamUsageEstimator.shallowSizeOf(futures));
+        int index = 0;
         for (List<T> partition : partitions) {
-            futures.add(supplyAsync(() -> {
+            allocateBytes.accept(NUM_BYTES_OBJECT_REF); // List pointer
+            allocateBytes.accept(RamUsageEstimator.sizeOfObject(partition)); // List contents
+            futures[index] = supplyAsync(() -> {
                 partition.sort(comparator);
+                allocateBytes.accept(sortOverheadBytes(partition.size()));
                 return partition;
-            }, executor));
+            }, executor);
+            index++;
         }
-        return CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]))
-            .thenApply(aVoid -> {
+        return CompletableFuture.allOf(futures)
+            .thenApply(_ -> {
+
+                // Re-using list to store sorted result.
+                // Saves NUM_BYTES_OBJECT_REF * list.size() bytes.
+                // Important: Parameter list is not used outside and thus can be mutated.
+                list.clear();
```

Comment:
Not sure reusing external parameter (and mutating it) is a good idea.

We do save memory a bit and currently mutation looks safe. 
However, it's roughly couple of Megabytes per million elements a we are saving only shallow size/pointers, 

Let's get back `ArrayList<T> result = new ArrayList<>(list.size())` and account for its shallow size?

Reply:
Update: facing same strange test failures. Even suspect that there can be another bug